### PR TITLE
Simplify sourcemap generation, and append `sourceMappingURL` section with objcopy

### PR DESCRIPTION
**Pull request by dschuff**
This PR simplifies the logic for handling the debug info required for source maps. In particular if debug info will not be needed later, it strips the DWARF earlier than previously (now in the finalize_wasm phase immediately after generating the source map and before finalization). It also uses objcopy rather than Binaryen to add the `sourceMappingURL` section that points to the source map. This is faster (in cases where Binaryen is not otherwise needed) and means that source map generation works even when using `-sERROR_ON_WASM_CHANGES_AFTER_LINK`.

Depends on https://reviews.llvm.org/D155535

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -3475,6 +3475,9 @@ def consume_arg_file():
         # set the emscripten debug level to 3 so that we do not remove that
         # debug info during link (during compile, this does not make a
         # difference).
+        else:
+          # Assume unknown non-integer levels are clang flags that emit dwarf.
+          settings.GENERATE_DWARF = 1
```
Comment (kripken) [inline emcc.py:3480]: Why is this needed?

**Review by kripken — COMMENTED** [review]
The code changes here look good, but the overall code is complex, so I would be less worried if this PR had no testing-related changes. Is that possible?

**Comment by aheejin** [comment]
@dschuff is not coming back for 2.5 weeks, and he wanted me to land this...

File: test/test_other.py [hunk 2]
```diff
@@ -2691,35 +2672,53 @@ def test_dwarf_sourcemap_names(self):
...
       (['--profiling-funcs'], False, False, True),
+      (['-O2', '--profiling-funcs'], False, False, True),
       (['-g'], True, False, True),
       (['-g3'], True, False, True),
-      (['-O2', '-g'], True, False, True),
+      (['-O1', '-g'], True, False, True),
+      (['-O3', '-g'], True, False, True),
       (['-gsplit-dwarf'], True, False, True),
+      # TODO: It seems odd that -gsource-map leaves behind a name section. Should it?
       (['-gsource-map'], False, True, True),
```
Comment (aheejin) [inline test/test_other.py:2690]: Not related to this PR, but why does `-gsource-map` generate the name section?

File: test/test_other.py [hunk 3]
```diff
@@ -2691,35 +2672,53 @@ def test_dwarf_sourcemap_names(self):
...
-      (['-g0', '-gsource-map'], False, True, True),
+      (['-g1', '-Oz', '-gsource-map'], False, True, True),
       # -g0 does not override -gsource-map but does remove name section. TODO: should it?
       (['-gsource-map', '-g0'], False, True, False),
+      (['--emit-symbol-map', '-gsource-map'], False, True, True),
+      (['--emit-symbol-map'], False, False, False),
+      (['--emit-symbol-map', '-Oz'], False, False, False),
```
Comment (aheejin) [inline test/test_other.py:2696]: Is `--emit-symbol-map` relevant to any of name section, DWARF, or source map generation? Adding it doesn't seem to change anything in these three options.

File: test/test_other.py [hunk 4]
```diff
@@ -2691,35 +2672,53 @@ def test_dwarf_sourcemap_names(self):
...
+      (['-gsource-map', '-sWASM_BIGINT', '-sERROR_ON_WASM_CHANGES_AFTER_LINK'], False, True, True),
     ]:
       print(flags, expect_dwarf, expect_sourcemap, expect_names)
       self.emcc(test_file(source_file), flags, js_file)
+      self.assertExists(js_file)
 
+      def notIn(member, container, msg=None):
+        if member in container:
+          print('%s unexpectedly found' % member)
+          self.fail()
```
Comment (aheejin) [inline test/test_other.py:2711]: This doesn't seem to be used anywhere

**Review by aheejin — COMMENTED** [review]
Code generally LGTM to me, even though it looks this contains several different changes and some test refactoring at the same time so I'm not sure which part is relevant to which tests. I'm also not sure which tests started to pass in this PR that hadn't passed before.

That being said, @dschuff is not coming back for 2.5 weeks, and he wanted me to land https://reviews.llvm.org/D155535 and this because it is requested by some user in our Wasm Users team chat. So @kripken, I think we can make whatever necessary changes and land this.

And I think the CI will pass after https://reviews.llvm.org/D155535 lands.

File: emscripten.py [hunk 5]
```diff
@@ -31,6 +32,9 @@
 from tools.shared import treat_as_user_export
 from tools.settings import settings
 
+sys.path.append(path_from_root('third_party'))
```
Comment (sbc100) [inline emscripten.py:35]: Is this line needed?  I would expect all the `tools` imports above to fail if it was

File: emscripten.py [hunk 6]
```diff
@@ -448,6 +452,15 @@ def get_metadata(infile, outfile, modify_wasm, args):
   return metadata
 
 
+def create_sourcemapping_url_file(infile, url):
+  'Create a file with the contents of the sourceMappingURL section'
+  filename = os.path.join(shared.get_emscripten_temp_dir(), infile + '.smu')
+  with open(filename, 'wb') as f:
```
Comment (sbc100) [inline emscripten.py:458]: How about `utils.write_binary(filename, leb128.u.encode(len(url) + url.encode('utf-8'))`.. maybe no better or worse?

File: emscripten.py [hunk 7]
```diff
@@ -448,6 +452,15 @@ def get_metadata(infile, outfile, modify_wasm, args):
   return metadata
 
 
+def create_sourcemapping_url_file(infile, url):
+  'Create a file with the contents of the sourceMappingURL section'
```
Comment (sbc100) [inline emscripten.py:456]: Use triple double quotes for doc strings.

File: emscripten.py [hunk 8]
```diff
@@ -448,6 +452,15 @@ def get_metadata(infile, outfile, modify_wasm, args):
...
+  filename = os.path.join(shared.get_emscripten_temp_dir(), infile + '.smu')
```
Comment (sbc100) [inline emscripten.py:457]: I think you can just do `shared.get_temp_files().get_file('.smu')`

File: test/test_other.py (hunk 2 above)
Comment (kripken) [reply test/test_other.py:2690]: I think that might be because we just set the internal debug level to "debug" (3 or 4) when generating source maps. We could probably avoid that with some work.

File: test/test_other.py (hunk 4 above)
Comment (aheejin) [reply test/test_other.py:2711]: Removed

File: test/test_other.py (hunk 3 above)
Comment (aheejin) [reply test/test_other.py:2696]: I see. Will add
```
# --emit-symbol-map should not affect the results
```
Above these three tests to be clear.

File: emscripten.py (hunk 5 above)
Comment (aheejin) [reply emscripten.py:35]: Why is this related to all the `tools` lines above?

This can be removed as you suggested, because `from tools import building` above eventually add `third_party` directory to `sys.path`:
https://github.com/emscripten-core/emscripten/blob/3cd9a380124bf17845f46819a895ec55a9e6ecf8/emscripten.py#L22
https://github.com/emscripten-core/emscripten/blob/3cd9a380124bf17845f46819a895ec55a9e6ecf8/tools/building.py#L23
https://github.com/emscripten-core/emscripten/blob/3cd9a380124bf17845f46819a895ec55a9e6ecf8/tools/webassembly.py#L18

But it is not easy to see this by just looking at `emscripten.py`. Not sure if we should depend on that `from tools import building` line is above?

File: emscripten.py (hunk 6 above)
Comment (aheejin) [reply emscripten.py:450]: Changed to
```py
utils.write_binary(filename, leb128.u.encode(len(url)) + url.encode('utf-8'))
```
(I think you meant this)

File: emscripten.py (hunk 7 above)
Comment (aheejin) [reply emscripten.py:448]: Done

File: emscripten.py (hunk 8 above)
Comment (aheejin) [reply emscripten.py:449]: It looks the URL file is supposed to exist after this file returns, but with `get_file`, is that possible? All uses of that methods tend to use the idiom of
```py
with shared.get_temp_files().get_file('.ext') as filename:
  do something
```
and the file is gone after this `with` block.

File: emcc.py [hunk 9]
```diff
@@ -3474,6 +3474,7 @@ def consume_arg_file():
             settings.SEPARATE_DWARF = requested_level.split('=')[1]
           else:
             settings.SEPARATE_DWARF = True
+          settings.GENERATE_DWARF = 1
```
Comment (aheejin) [inline emcc.py:3477]: Without this, `SEPARATE_DWARF` is not recognized as generating DWARF and other related logic in this PR doesn't work

File: test/code_size/hello_world_wasm.json [hunk 10]
```diff
@@ -3,8 +3,8 @@
   "a.html.gz": 431,
   "a.js": 292,
   "a.js.gz": 252,
-  "a.wasm": 104,
+  "a.wasm": 103,
```
Comment (aheejin) [inline test/code_size/hello_world_wasm.json:6]: Size has shrunk by 1 byte because `wasm-opt` doesn't strip debug/producer sections anymore, which may sound strange. The saving comes from the code section, and what I suspect is, after generating and optimizing stack IR passes, an empty function does not contain any instruction, but before we used to strip debug and producer sections with `wasm-opt` and a `nop` is added in the process of that. This does not happen anymore because we strip those sections with `llvm-objcopy`. I don't now why stripping those sections cause a `nop` to be added. I dug a little, but I'm not sure it's worth more investigation.. It's 1 byte, and it got better after all.

File: emcc.py (hunk 1 above)
Comment (aheejin) [reply emcc.py:3488]: Not sure why it is needed. Will remove it for now. @dschuff, feel free to add it back later if necessary.

Comment (aheejin) [reply emcc.py:3488]: I don't know why this is needed either, and not sure what to do with this. (And I don't think tests here cover this) Leave it or delete it?

File: test/test_other.py [hunk 11]
```diff
@@ -2569,28 +2569,9 @@ def test_emcc_debug_files(self):
           if debug is None:
             self.assertFalse(os.path.exists(self.canonical_temp_dir))
           elif debug == '1':
-            self.assertExists(os.path.join(self.canonical_temp_dir, 'emcc-3-original.js'))
+            self.assertExists(os.path.join(self.canonical_temp_dir, 'emcc-4-original.js'))
```
Comment (aheejin) [inline test/test_other.py:2572]: This is because the order of stripping changed:

Before:
```console
emcc-0-base.wasm 
emcc-1-wasm-emscripten-finalize.wasm
emcc-2-post_finalize.wasm
emcc-3-original.js
emcc-4-strip.wasm
...
```
After:
```console
emcc-0-base.wasm
emcc-1-strip.wasm
emcc-2-wasm-emscripten-finalize.wasm
emcc-3-post_finalize.wasm
emcc-4-original.js
...
```

**Comment by aheejin** [comment]
The CI keeps failing with this message:
```
Checkout is not clean.  See above for list of dirty/untracked files.

Exited with code exit status 1
```

Any idea on how to fix this?

File: emscripten.py (hunk 8 above)
Comment (sbc100) [reply emscripten.py:449]: Ah, in that case maybe `shared.get_temp_files().get(...)` is what you want then.

File: tools/building.py [hunk 12]
```diff
@@ -1217,7 +1218,7 @@ def run_binaryen_command(tool, infile, outfile=None, args=None, debug=False, std
       # in are not enough to see what went wrong)
       if settings.LEGALIZE_JS_FFI:
         extra += '\nnote: to disable int64 legalization (which requires changes after link) use -sWASM_BIGINT'
-      if settings.OPT_LEVEL > 0:
+      if settings.OPT_LEVEL > 1:
```
Comment (sbc100) [inline tools/building.py:1221]: Is this just an existing bug in the condition here?

File: tools/building.py [hunk 13]
```diff
@@ -920,6 +920,7 @@ def wasm2js(js_file, wasm_file, opt_level, minify_whitespace, use_closure_compil
 
 
 def strip(infile, outfile, debug=False, sections=None):
+  'Strip DWARF and/or other specified sections from a wasm file'
```
Comment (sbc100) [inline tools/building.py:923]: Use triple double quotes for docstrings.

File: test/test_other.py [hunk 14]
```diff
@@ -2569,28 +2569,9 @@ def test_emcc_debug_files(self):
...
-        (['-O2', '-g'], True),
-      ]:
-      print(args, expect_debug)
-      err = self.run_process([EMCC, '-v', test_file('hello_world.c')] + args, stdout=PIPE, stderr=PIPE).stderr
-      lines = err.splitlines()
-      finalize = [l for l in lines if 'wasm-emscripten-finalize' in l][0]
-      if expect_debug:
-        self.assertIn(' -g ', finalize)
-      else:
-        self.assertNotIn(' -g ', finalize)
```
Comment (sbc100) [inline test/test_other.py:2593]: Why is the whole test being removed?

File: tools/building.py (hunk 12 above)
Comment (aheejin) [reply tools/building.py:1221]: I.. think so?

File: emscripten.py (hunk 8 above)
Comment (aheejin) [reply emscripten.py:449]: Done

File: tools/building.py (hunk 13 above)
Comment (aheejin) [reply tools/building.py:923]: Done

File: test/test_other.py (hunk 14 above)
Comment (aheejin) [reply test/test_other.py:2593]: I think it was merged into `test_dwarf_sourcemap_names`.

File: emscripten.py (hunk 8 above)
Comment (aheejin) [reply emscripten.py]: Hmm, come to think of it, the original code assumes a specific filename, not a generated random name. We can probably use `shared.get_temp_files().get_dir()`, but I guess it isn't very different from what @dschuff wrote. I'll revert it.

**Comment by aheejin** [comment]
> The CI keeps failing with this message:
> 
> ```
> Generating XML reports...
> Running 'git status --short'
> 
> m test/third_party/wasi-test-suite
> 
> Checkout is not clean.  See above for list of dirty/untracked files.
> 
> Exited with code exit status 1
> ```
> 
> Any idea on how to fix this?

This was because, when `--post-link` was given, which apparently meant running only the post-link phase, running `emcc` changed the _input_ which is supposed to be untouched. https://github.com/emscripten-core/emscripten/pull/19890/commits/bff1ceddeb359b9ecca4bb3ba5463c5d922607a7 fixes this problem.

In [`final_wasm`](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emscripten.py#L443-L538), where the newly added stripping code is located, when we [compile and link](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emcc.py#L226), `infile` and `outfile` parameters contain the same, newly-generated output file. But when `--post-link` is given, we [only run the post-link](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emcc.py#L225) phase, which [is the case](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/test/test_other.py#L13371) for the [WASI tests](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/test/test_other.py#L13379-L13433), `infile` points to the input wasm file and `outfile` points to the output file. And the previous code tried to strip sections from `infile`.

Now all CI passes.

File: tools/building.py (hunk 12 above)
Comment (dschuff) [reply tools/building.py:1221]: Yes, it is. (it was noted in the wasm-users discussion that reported the main bug this PR fixes, i.e. the fact that source map generation doesn't work with ERROR_ON_WASM_CHANGES_AFTER_LINK)

File: emcc.py (hunk 1 above)
Comment (dschuff) [reply emcc.py:3495]: I think this is here to cover flags such as `-gdwarf-5` which causes DWARF to be emitted and also causes DWARF5 to be used. Maybe I should add a test for that.

Comment (dschuff) [reply emcc.py:3495]: Oh actually the comment just above this `else` clauses also describes this use case. It's weird that they are separated, I'll consolidate them.

**Comment by dschuff** [comment]
> This was because, when `--post-link` was given, which apparently meant running only the post-link phase, running `emcc` changed the _input_ which was supposed to be untouched. [bff1ced](https://github.com/emscripten-core/emscripten/commit/bff1ceddeb359b9ecca4bb3ba5463c5d922607a7) fixes this problem.
> 
> In [`final_wasm`](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emscripten.py#L443-L538), where the newly added stripping code is located, when we [compile and link](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emcc.py#L226), `infile` and `outfile` parameters contain the same, newly-generated wasm file. But when `--post-link` is given, we [only run the post-link](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emcc.py#L225) phase, which [is the case](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/test/test_other.py#L13371) for the [WASI tests](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/test/test_other.py#L13379-L13433), `infile` points to the input wasm file and `outfile` points to the output file. And the previous code tried to strip sections from `infile`, stripping the producer section from wasm files under `test/third_party/wasi-test-suites`.
> 
> Now all CI passes.

Moving stripping after the `wasm-emscripten-finalize` run is slightly a bummer since stripping before finalization should in theory result in finalization being a bit faster. I tried changing it so that the strip output is a different file from the input (i.e. the stripping is not in-place). WDYT?

File: emscripten.py [hunk 15]
```diff
@@ -498,11 +504,38 @@ def finalize_wasm(infile, outfile, memfile, js_syms):
...
+    # For sections we no longer need, strip now to speed subsequent passes
+    post_strip_file = shared.get_temp_files().get('.wasm').name
+    building.save_intermediate(infile, 'strip.wasm')
+    sections = ['producers'] if not settings.EMIT_PRODUCERS_SECTION else []
+    building.strip(infile, post_strip_file, debug=not settings.GENERATE_DWARF,
+                   sections=sections)
+
+  metadata = get_metadata(post_strip_file, outfile, modify_wasm, args)
   if modify_wasm:
     building.save_intermediate(infile, 'post_finalize.wasm')
```
Comment (aheejin) [inline emscripten.py:535]:
```suggestion
    building.save_intermediate(post_strip_file, 'post_finalize.wasm')
```
Shouldn't this be changed too?

**Comment by aheejin** [comment]
> > This was because, when `--post-link` was given, which apparently meant running only the post-link phase, running `emcc` changed the _input_ which was supposed to be untouched. [bff1ced](https://github.com/emscripten-core/emscripten/commit/bff1ceddeb359b9ecca4bb3ba5463c5d922607a7) fixes this problem.
> > In [`final_wasm`](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emscripten.py#L443-L538), where the newly added stripping code is located, when we [compile and link](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emcc.py#L226), `infile` and `outfile` parameters contain the same, newly-generated wasm file. But when `--post-link` is given, we [only run the post-link](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/emcc.py#L225) phase, which [is the case](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/test/test_other.py#L13371) for the [WASI tests](https://github.com/emscripten-core/emscripten/blob/6b6f7e7ed4aa8c1f2232443403fc8491056b01f6/test/test_other.py#L13379-L13433), `infile` points to the input wasm file and `outfile` points to the output file. And the previous code tried to strip sections from `infile`, stripping the producer section from wasm files under `test/third_party/wasi-test-suites`.
> > Now all CI passes.
> 
> Moving stripping after the `wasm-emscripten-finalize` run is slightly a bummer since stripping before finalization should in theory result in finalization being a bit faster. I tried changing it so that the strip output is a different file from the input (i.e. the stripping is not in-place). WDYT?

Looks fine to me w/ a question.

File: emscripten.py (hunk 15 above)
Comment (dschuff) [reply emscripten.py]: Yes. But see below

**Comment by dschuff** [comment]
OK, something isn't quite right with that last change, I've reverted it. I think we can land this as-is and optimize it in a separate PR.

**Comment by aheejin** [comment]
> OK, something isn't quite right with that last change, I've reverted it. I think we can land this as-is and optimize it in a separate PR.

I re-reverted your reverting and made some changes (https://github.com/emscripten-core/emscripten/pull/19890/commits/94111f25ee3e66c5d07ab2651bdae8c2d2d7898a), and it seems to work. I removed `post_strip_file` variable, and made the copying `infile` to `outfile` in case they are different be done up front, and used `outfile` for the rest of the code.

**Comment by aheejin** [comment]
Gentle ping 😀 Can we land this?

File: emscripten.py (hunk 8 above)
Comment (sbc100) [reply emscripten.py:449]: I still think doesn't look right.  Why can't `shared.get_temp_files().get(...)` work?   I don't think the calling code should care what the filename is. 

Also, how about calling this `.sec` (for section?)?  Or even `.bin`?

File: test/test_other.py [hunk 16]
```diff
@@ -2691,35 +2672,49 @@ def test_dwarf_sourcemap_names(self):
       ([], False, False, False),
       (['-g0'], False, False, False),
       (['-g1'], False, False, False),
+      (['-g1', '-O2'], False, False, False),
```
Comment (kripken) [inline test/test_other.py:2675]: Please add a line with plain `-O2` and nothing else (like the old test had) - just to be safe here.

**Review by kripken — APPROVED** [review]
lgtm % last comment + @sbc100 's comments

File: emscripten.py (hunk 8 above)
Comment (aheejin) [reply emscripten.py:449]: I think it can work if we just delete `create_sourcemapping_url_file` and inline its contents. That way we can use `with` clause and don't need to delete the file afterwards. Currently we just leave the temporary file. I also changed its extension to `.bin` as suggested. So I deleted `create_sourcemapping_url_file` and changed its callsite to:
```py
    else:
      # Otherwise use objcopy. This avoids re-encoding the file (thus
      # preserving DWARF) and is faster.

      # Create a file with the contents of the sourceMappingURL section    
      with shared.get_temp_files().get_file('.bin') as url_file:
        utils.write_binary(url_file,
                           leb128.u.encode(len(base_url)) + base_url.encode('utf-8'))
        cmd = [shared.LLVM_OBJCOPY,
               '--add-section',
               'sourceMappingURL=' + url_file,
               infile]
        shared.check_call(cmd)
```

File: test/test_other.py (hunk 16 above)
Comment (aheejin) [reply test/test_other.py:2675]: Done

File: emscripten.py [hunk 17]
```diff
@@ -498,13 +497,43 @@ def finalize_wasm(infile, outfile, memfile, js_syms):
...
+      # Otherwise use objcopy. This avoids re-encoding the file (thus
+      # preserving DWARF) and is faster.
+
+      # Create a file with the contents of the sourceMappingURL section
+      with shared.get_temp_files().get_file('.bin') as url_file:
+        utils.write_binary(url_file,
+                           leb128.u.encode(len(base_url)) + base_url.encode('utf-8'))
+        cmd = [shared.LLVM_OBJCOPY,
+               '--add-section',
+               'sourceMappingURL=' + url_file,
```
Comment (kripken) [inline emscripten.py:520]: It looks like this puts the location of a temporary file in this section? But I think we want it to point to the final location of the source map?

(In fact I'm confused as to why we want a temp file here?)

Comment (sbc100) [reply emscripten.py:520]: The location of the file is not relevant.  Its just the `--add-section'` requires a file the contains the section contents.

Comment (kripken) [reply emscripten.py:520]: Thanks, I see, it just reads the contents... I misunderstood then.

**Review by kripken — APPROVED** [review]

File: emscripten.py (hunk 17 above)
Comment (aheejin) [reply emscripten.py:520]: I was confused by this too at first. It turns out, `llvm-objcopy`'s `--add-section` option takes a file name whose contents will be put in the requested section. From [`llvm-objcopy`'s manual page](https://llvm.org/docs/CommandGuide/llvm-objcopy.html):
```
--add-section <section=file>
Add a section named <section> with the contents of <file> to the output. For ELF objects the section will be of type SHT_NOTE, if the name starts with “.note”. Otherwise, it will have type SHT_PROGBITS. Can be specified multiple times to add multiple sections.

For MachO objects, <section> must be formatted as <segment name>,<section name>.
```
So we have to create a temp file and write the URL of the map file into this temp file.

Comment (aheejin) [reply emscripten.py:520]: Oops, I laced with Sam.

**Comment by aheejin** [comment]
@sbc100 Do you have more concerns?

**Review by sbc100 — APPROVED** [review]
I haven't groked all the details, but lgtm if others have.

**Comment by dschuff** [comment]
Thanks for the help on this @aheejin
